keps/sig-node/4191-split-image-filesystem/README.md

```diff
@@ -0,0 +1,899 @@
+Generally each filesystem will need to be tracked as part of the Stats Provider and that should be used to populate the Summary API.  
+The eviction manager would need to be aware of each filesystem that is added.
+
+Should CAdvisor monitor total capacity of each filesystem or should that be returned in CRI?
+
+Adding a new API for ImageFsStats that returns an array of both images and containers would allow for possible extension for CRI implementations to report file system totals and usage.
+
+Another extension can be multiple filesystems for images and containers.  
+The API allows for a list of filesystem usage per images and containers but there has been no work done to support
+this in other container runtimes.  
+
+### Risks and Mitigations
+
+By splitting the filesystem we allow more cases than what we currently support in Kubelet.
+To avoid bugs, we will validate on cases we don't currently support in Kubelet and return an error.
+
+The following cases will be validated and we will return an error if container runtime is set up for this:
+
+- Read-only layer and writable layer in separate locations on a separate disk from root filesystem.
+- Read-Only layer on root filesystem while Writeable layer on image filesystem.
+- Read-Only, Writable and root are all on separate disks.
+
+## Design Details
+
+### CRI
+
+CRIO and containerd return a single element in this case and Kubelet does not assume that there are multiple values in this array.  Regardless, we add an array to ImageFsInfoResponse.
+
+```golang
+// ImageService defines the public APIs for managing images.
+service ImageService {
+…
+    rpc ImageFsInfo(ImageFsInfoRequest) returns (ImageFsInfoResponse) {}
+}
+
+message ImageFsInfoResponse {
+    // Information of image filesystem(s).
+    repeated FilesystemUsage image_filesystems = 1;
+    // Information of container filesystem(s).
+    // This is an optional field if container and image
+    // storage are separated.
+    // Default will be to return this as empty.
+    repeated FilesystemUsage container_filesystems = 2;
+}
+
+// FilesystemUsage provides the filesystem usage information.
+message FilesystemUsage {
+    // Timestamp in nanoseconds at which the information were collected. Must be > 0.
+    int64 timestamp = 1;
+    // The unique identifier of the filesystem.
+    FilesystemIdentifier fs_id = 2;
+    // UsedBytes represents the bytes used for images on the filesystem.
+    // This may differ from the total bytes used on the filesystem and may not
+    // equal CapacityBytes - AvailableBytes.
+    UInt64Value used_bytes = 3;
+    // InodesUsed represents the inodes used by the images.
+    // This may not equal InodesCapacity - InodesAvailable because the underlying
+    // filesystem may also be used for purposes other than storing images.
+    UInt64Value inodes_used = 4;
+}
+```
+
+CRIO returns a unique identifier for images and containers so Kubelet can ask CRI if the objects are stored on separate disks.  
```

Comment:
I don't know if we should reference one of the CRI impl's impelementation details without the other. Maybe make it more passive voice like:

```
It is expected of the CRI implementation to return a unique identifier for images and containers so the Kubelet can ask CRI if the objects are stored on separate disks.
```

Reply:
done.